### ISSUE TITLE
Rule proposal: Make TS 3.7 changes affect eslint control flow

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#1022]
https://github.com/microsoft/TypeScript/pull/32695 added 2 structures that affect control flow:

1. Functions that return `never`:

	```ts
	declare function fail(): never;
	
	function test() {
		switch (null) {
			case 0:
				fail();
				// no-fallthrough shouldn't trigger
			case 1:
		}
	}
	```

2. Exhaustive switch cases (before it worked only for function return value):

	```ts
	function test() {
		switch (null) {
			case 0:
				declare const foo: 'foo' | 'bar';
				switch (foo) {
					case 'foo': return;
					case 'bar': return;
				}
				// no-fallthrough shouldn't trigger
			case 1:
		}
	}
	```

Related: [process-exit-as-throw](https://github.com/mysticatea/eslint-plugin-node/blob/master/lib/rules/process-exit-as-throw.js) from eslint-plugin-node

**Comment by bradzacher** [comment]
There's a much larger piece of work here than you might think.

We need Typescript to expose its control flow logic. Otherwise we have to reimplement it all ourselves.

**Comment by ark120202** [comment]
For never-returning functions it shouldn't be too hard, the general logic is the same as in process-exit-as-throw, but instead of just checking for `process.exit` it would check for return type of called expression.

As for exhaustive switch it's indeed more complex than I have expected, I have assumed there was some way to get node links from type checker, but looks like it's not exposed even in internal APIs. Well, it doesn't look like a common case anyway, so leaving it probably wouldn't be a huge issue.

**Comment by JamesHenry** [comment]
I believe the resolution here is that there is no action that can be taken in this project because of the lack of underlying APIs, so closing for now.

Thanks all 🙏
